source/common/common/trie_lookup_table.h

```diff
@@ -0,0 +1,167 @@
+#pragma once
+
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+
+/**
+ * A trie used for faster lookup with lookup time at most equal to the size of the key.
+ *
+ * Type of Value must be empty-constructible and moveable, e.g. smart pointers and POD types.
+ */
+template <class Value> class TrieLookupTable {
+  // A TrieNode aims to be a good balance of performant and
+  // space-efficient, by allocating a vector the size of the range of children
+  // the node contains. This should be good for most use-cases.
+  //
+  // For example, a node with children 'a' and 'z' will contain a vector of
+  // size 26, containing two values and 24 nulls. A node with only one
+  // child will contain a vector of size 1. A node with no children will
+  // contain an empty vector.
+  //
+  // Compared to allocating 256 entries for every node, this makes insertions
+  // a little bit inefficient (especially insertions in reverse order), but
+  // trie lookups remain O(length-of-longest-matching-prefix) with just a
+  // couple of very cheap operations extra per step.
+  //
+  // By size, having 256 entries for every node makes each node's overhead
+  // (excluding values) consume 8KB; even a trie containing only a single
+  // prefix "foobar" consumes 56KB.
+  // Using ranged vectors like this makes a single prefix "foobar" consume
+  // less than 20 bytes per node, for a total of less than 0.14KB.
+  //
+  // Using indices instead of pointers helps keep the bulk of the data
+  // localized, and prevents recursive deletion which can provoke a stack
+  // overflow.
+  struct TrieNode {
+    Value value_{};
+    // -1 in any index where there is not a child.
+    std::vector<int32_t> children_;
```

Comment:
comment that these are indexes into TrieLookupTable::nodes_

Reply:
Done.